Fixes #3073 - Fix mismatched and missing Reboot/Reset power actions

spesnova717 opened the pull request:
Bug [#3073](https://projects.theforeman.org/issues/3073?tab=history) is caused by 3 things:
1. Foreman core doesn't call BMC proxy's API correctly
   - Changing power state to 'Reboot' via Web GUI calls BMC proxy's 'soft' power action
   - Changing power state to 'Reset' via Web GUI calls BMC proxy's 'cycle' power action
2. Foreman BMC proxy doesn't support reboot at all BMC provider
3. Foreman BMC proxy doesn't support reset(powerreset) at Redfish provider
 
This PR fixes first item.

## Modifications in this PR

### (1) Removal of Incorrect Power Action Mappings

In `bmc.rb`, the `:reboot` and `:reset` actions were incorrectly mapped to `soft` and `cycle`, respectively.  
These incorrect mappings are removed to eliminate the discrepancy between the Web UI representation and the actual behavior.

### (2) Implement Reboot and Reset Actions

In the `power` method of `proxy_api/bmc.rb`, the `reboot` and `reset` actions were not previously implemented.  
These actions are implemented, enabling Foreman Core to send `reboot` and `reset` power control requests to the Smart Proxy.

spesnova717 committed:
Fixes #3073 - Fix mismatched and missing Reboot/Reset power actions

Bug #3073 is caused by 3 things:
1. Foreman core doesn't call BMC proxy's API correctly
   - Changing power state to 'Reboot' via Web GUI calls BMC proxy's 'soft'
     power action
   - Changing power state to 'Reset' via Web GUI calls BMC proxy's 'cycle'
     power action
2. Foreman BMC proxy doesn't support reboot at all BMC provider
3. Foreman BMC proxy doesn't support reset(powerreset) at Redfish provider

This PR fixes first item.

For backward compatibility, power_action_v2 capability will be
introduced to smart proxy at PRs related to #38498. And this commit
introduces logic to determine if smart proxy supports that capability.

Co-Authored-By: Vanou Ishii <ishii.vanou@fujitsu.com>

spesnova717 commented:
@lzap san

The content of the closed PR [#9082](https://github.com/theforeman/foreman/pull/9082) is addressed in the new PR [#10561](https://github.com/theforeman/foreman/pull/10561).
As this change also requires updates on the smart-proxy side, we created the following related ticket & PRs: 
- Ticket
  * [Bug #38498 Wrong/missing implementation of BMC proxy power operation](https://projects.theforeman.org/issues/38498)

- PR
  * [#916](https://github.com/theforeman/smart-proxy/pull/916)
  * [#917](https://github.com/theforeman/smart-proxy/pull/917)
 
All three PRs (one in Foreman and two in smart-proxy) are **ready for review**.  
Your feedback would be greatly appreciated.

ekohl reviewed (commented):
I'm thinking about compatibility here. Leaving this on the PR here, but it will apply to the Smart Proxy too.

We support running Foreman with older Smart Proxies so I'd like to detect if the proxy actually supports these (in other words, runs with or without your bug fixes). There's a mechanism with capabilities and in https://theforeman.org/2019/04/smart-proxy-capabilities-explained.html I've tried to describe how it works.

On the Smart Proxy side it will be easy: simply add a `capability` statement to signal it supports these. On the Foreman side you can use `proxy.has_capability?(:BMC, :my_capability)` to detect it.

Do you think you could take a look at this?

lzap commented:
I will leave the review on my colleagues. Thank you so much for your contribution!

spesnova717 commented:
@ekohl san

Thank you for your comment.
I'll proceed with the necessary modifications.
I'd appreciate it if you clarify the following points:
 
### 1. Support policy of version mixing
 
According to Foreman's [release policy](https://projects.theforeman.org/projects/foreman/wiki/Release_Management#Minor-releases), if latest Foreman release is `3.x`, `3.x` and `3.x-1` is supported.
 
Then, following combination of Foreman core & smart proxy is also supported, is this correct?
 
- Foreman core `3.x` ― Smart Proxy `3.x`
- Foreman core `3.x` ― Smart Proxy `3.x-1`
 
But this isn't supported, right?
 
- Foreman core `3.x-1` ― Smart Proxy `3.x`
 
 
### 2. Should capability continue to remain in code, even after all supported version of Foreman core&proxy apply our modification?
 
We'll follow your advice and will add capability (e.g., `proxy.has_capability?(:BMC, :<my_capability>)`) to Foreman.
 
If our modification is included at Foreman version `3.x` and when `3.x+2` is released, all supported version of Foreman core and smart proxy (`3.x+1` & `3.x+2`) includes our modification.
At this point, can we remove capability & related code (e.g. `if something.has_capability?` line) from source code of Foreman core and smart proxy? Or should we let that capability related code remains in code forever?
 
 
### 3. What is expected behavior regarding compatibility?
 
At current Foreman (before our PR is merged), pushing `Reboot` at GUI leads to `soft` & pushing `Reset` at GUI leads to `cycle`.
 
Regarding compatibility, expected behavior is following, right?
 
If our modification is merged at `3.x`, Foreman core (version `3.x`) behavior:
* Against smart proxy which includes our modification (version `3.x`, with `<my_capability>`)
  - pushing `Reboot` at GUI leads to `Reboot`
  - pushing `Reset` at GUI leads to `Reset` (powerreset)
* Against smart proxy which doesn't have our modification (version `3.x-1`, without `<my_capability>`)
  - pushing `Reboot` at GUI leads to `soft`
  - pushing `Reset` at GUI leads to `cycle`

Thank you.

ekohl commented:
> ### 1. Support policy of version mixing
> 
> According to Foreman's [release policy](https://projects.theforeman.org/projects/foreman/wiki/Release_Management#Minor-releases), if latest Foreman release is `3.x`, `3.x` and `3.x-1` is supported.
> 
> Then, following combination of Foreman core & smart proxy is also supported, is this correct?
> 
>     * Foreman core `3.x` ― Smart Proxy `3.x`
> 
>     * Foreman core `3.x` ― Smart Proxy `3.x-1`
> 
> 
> But this isn't supported, right?
> 
>     * Foreman core `3.x-1` ― Smart Proxy `3.x`

That is correct: we only support Foreman >= Smart Proxy.

> 2. Should capability continue to remain in code, even after all supported version of Foreman core&proxy apply our modification?

We have a downstream called Red Hat Satellite that has a wider support cycle. There can be 4 versions difference. That's why it's probably best to keep it for some time. I'd consider it sufficient to leave a code comment in Foreman that's something like Smart Proxy x.y introduced capability C so in the future we don't need to dig too deep.

> If our modification is merged at `3.x`, Foreman core (version `3.x`) behavior:
> 
>     * Against smart proxy which includes our modification (version `3.x`, with `<my_capability>`)
>       
>       * pushing `Reboot` at GUI leads to `Reboot`
>       * pushing `Reset` at GUI leads to `Reset` (powerreset)
> 
>     * Against smart proxy which doesn't have our modification (version `3.x-1`, without `<my_capability>`)
>       
>       * pushing `Reboot` at GUI leads to `soft`
>       * pushing `Reset` at GUI leads to `cycle`

That sounds good. I'd also accept a clear error message that the requested operation is not supported. I'll admit I'm not too familiar with the exact BMC implementation(s) to say what's best.

spesnova717 commented:
Thank you for your response.
All clear now.
Thank you.

spesnova717 commented:
@ekohl 

Thank you for the feedback.
Yes, I understand the importance of maintaining compatibility with older Smart Proxies.
To address this:
* **On the Smart Proxy side**, I’ve added the `poweraction_fix_3_16` capability in `bmc_plugin.rb`. This will be included as part of [Fixes #38498 - Add support for Reset (powerreset) via Redfish provider](https://github.com/theforeman/smart-proxy/pull/916).
* **On the Foreman side**, I implemented capability detection using `has_capability?(:BMC, :poweraction_fix_3_16)` within `PowerManager::BMC`. Based on the presence of this capability, the `reboot` and `reset` actions are conditionally mapped to `do_reboot` / `do_reset` or fallback to `soft` / `cycle`.

This ensures proper behavior with both updated and legacy proxies, maintaining backward compatibility while enabling the new functionality where supported.
Please let me know if you have further suggestions or if anything needs to be adjusted.

The CI error originates from the `test:katello` job and is not caused by the changes in this PR. Similar failures are currently occurring in other PRs as well:
https://github.com/theforeman/foreman/actions/runs/16010470304

Thank you for your support.

spesnova717 commented:
Hi @ekohl, @lzap

This is a gentle ping for review on the following PRs:
[PR #10561](https://github.com/theforeman/foreman/pull/10561),
[PR #916](https://github.com/theforeman/smart-proxy/pull/916),
[PR #917](https://github.com/theforeman/smart-proxy/pull/917)

These address mismatched and missing Reboot / Reset BMC power actions.

Although the CI currently shows ❌, the failure appears to be unrelated to this PR.
Specifically, the error is a `Timeout::Error: execution expired` in a multiselect-related JS integration test,
which is not affected by the changes in this PR.
It seems likely that this is a temporary CI-side issue.

If someone reviews it, it would be much appreciated.

If there’s anything blocking the review or merge process,
please let me know — I would like to help you if possible.

I look forward to hearing from you.

evgeni commented:
Hi,

I am currently looking into this topic together with @stejskalleos and must admit I am a bit lost on vocabulary, so hoped you can help me a bit here.

Looking at our [current BMC API in the Smart Proxy](https://github.com/theforeman/smart-proxy/blob/3d74a04997bb222fb488cb108eeb4fadd39235cd/modules/bmc/bmc_api.rb#L163-L189), we support two "reset" options: `cycle` and `reset`.
In Foreman, we currently map the "Reset" action to `cycle`, and "Reboot" is wrongly mapped to `soft` (shutdown).

What I am lost is what these words supposed to mean.
I assume (!) "Reboot" means a soft reboot (like when you type `reboot` in a root shell, or press `Ctrl-Alt-Delete`) and a "Reset" is a hard reset (like when you cut power to the device)?

Skimming the IPMI 2.0 spec, especially chapter "Chassis Control Command", I see that there is an command of "power cycle" and one for "hard reset", with the former being marked as optional. Both are exposed in our [IPMI implementation on the Smart Proxy](https://github.com/theforeman/smart-proxy/blob/3d74a04997bb222fb488cb108eeb4fadd39235cd/modules/bmc/ipmi.rb#L51-L59).

Now, looking at Redfish Resource and Schema Guide, especially ResetType property, it can have `ForceRestart`, `GracefulRestart` and `PowerCycle`. I guess `GracefulRestart` is something that IPMI has no matching command for? Our [Redfish implementation in the Smart Proxy](https://github.com/theforeman/smart-proxy/blob/3d74a04997bb222fb488cb108eeb4fadd39235cd/modules/bmc/redfish.rb#L73-L75) only does `ForceRestart` but exposes it as `powercycle` (???)

We also have "Shell" and "SSH" pseudo-BMCs, that only expose `powercycle` by calling `shutdown -r`

Do we agree on the meaning of "Reboot" (soft) and "Reset" (hard)?
Then I think the mapping could to be:
* IPMI: Reboot: `reset` (unless there is a way to implement soft reboot, which I don't see), Reset: `reset`
* Redfish: Reboot: `GracefulRestart`, Reset: `ForceRestart` or `PowerCycle` -- not sure which one is better

And based on that, we could define the new proxy capability as `powerreboot` and let the implementations call what they think is needed there?

Edit: after talking to friends from OpenStack - we could do a "soft off, on" sequence for the soft reboot on IPMI?

vanou commented:
Hello, @evgeni and @stejskalleos. Thanks for review this PR.

(Edit: Sorry, I could not afford to write down root cause/detailed explanation in text format at the time when I post this comment. So I added/modified following)

**--- Addition/modification Start ---------**
Actually, this is bit complex problem. As you said, followings are bit confusing area

* What does each term in IPMI specification & Redfish specification mean (e.g. Reboot, Reset, Powercycle, etc.)
* Complex relationship (and misbehavior) in following things
  * Foreman core behavior & how Foreman core calls BMC API at Smart Proxy
  * BMC API at Smart Proxy
  * IPMI/Redfish implementation at Smart Proxy

So I created following table with hope that it explains the situation well.
Please ask any question if this is not helpful.

In brief, this is not only about vocabulary, it is about vocabulary & wrong implementation of Foreman core/Proxy, I think. So we named capability as `poweraction_fix_3_17`.

> What I am lost is what these words supposed to mean.
> I assume (!) "Reboot" means a soft reboot (like when you type reboot in a root shell, or press Ctrl-Alt-Delete) and a "Reset" is a hard reset (like when you cut power to the device)?

I agree with you about what "Reboot" and "Reset" means.

**About Reboot at IPMI Provider**
My opinion for Reboot operation at IPMI provider is "not implement Reboot against IPMI provider" (at least) in this PRs. 

"soft off, on" sequence is good approach to Reboot on IPMI, but it is bit complex & expensive operation:
* We need to wait until/detect node is completely shutdown before triggering 'on'. Or if it wrongly trigger 'on' during node is shutting down, instruction of 'on' does not work in expected way.)
* To detect node is completely shutdown, BMC Smart Proxy needs to periodically communicate with server's BMC. If we trigger 'reboot' against many nodes through Foreman power operation, BMC Smart Proxy's tapping against many nodes affects network traffic quality.

If we decide to implement Reboot functionality against IPMI, it should be good to make another PR because above requirements for Reboot on IPMI is bit complex.

**--- Addition/modification End ---------**

|Power action which Foreman UI supports[1]|Corresponding IPMI chassis control command[2]|Corresponding Redfish ResetType[3]|What power action *current* Foreman core determines to call[4]|What *current* Foreman core actually calls against BMC Proxy API[5]|What *current* BMC proxy API listening on & what provder method is called[6]|What method *current* IPMI provider implements[7]|What method *current* Redfish provider implements[8]|
|---|---|---|---|---|---|---|---|
|Start|1h<br>(power up)|On|`on`|`on`|`on` -> `@bmc.poweron`|`poweron`|`poweron`|
|Stop|0h<br>(power down)|ForceOff|`off`|`off`|`off` -> `@bmc.poweroff`|`poweroff`|`poweroff`|
|Poweroff|0h<br>(power down)|ForceOff|`off`|`off`|`off` -> `@bmc.poweroff`|`poweroff`|`poweroff`|
|Reboot|*none*<br>(there isn't corresponding power operation in IPMI)|GracefullRestart|`soft`<br>(wrong, should be `reboot`)<br>**Fixed by [PR10561]**|`soft`<br>(wrong, should be `reboot`)<br>**Fixed by [PR10561]**|*none*<br>(not implemented, should be implemented: `reboot` -> `@bmc.powerreboot`)<br>**Fixed by [PR917]**|*none*<br>(because IPMI doesn't support this, corresponding method should not be implemented)|*none*<br>(not implemented, corresponding method `powerreboot` should be implemeted)<br>**Fixed by [PR917]**|
|Reset|3h<br>(hard reset)|ForceRestart|`cycle`<br>(wrong, should be `reset`)<br>**Fixed by [PR10561]**|`cycle`<br>(wrong, should be `reset`)<br>**Fixed by [PR10561]**|`reset` -> `@bmc.powerreset`|`powerreset`|*none*<br>(not implemented, corresponding method `powerreset` should be implemented)<br>**Fixed by [PR916]**|
|Soft|5h<br>(initiate a soft-shutdown)|GracefulShutdown|`soft`<br>(via [d])|`soft`|`soft` -> `@bmc.poweroff(true)`|`poweroff(true)`|`poweroff(true)`|
|Cycle|2h<br>(power cycle)|PowerCycle|`cycle`<br>(via [d])|`cycle`|`cycle` -> `@bmc.powercycle`|`powercycle`|`powercycle`<br>(`powercycle` method is implemented, but internal implementation is **wrong**: current internal implementation calls `ForceRestart` but should call `PowerCycle`)<br>**Fixed by [PR919]**|

* [1] What power action Foreman can trigger is defined at [a] & [b]
  * [a] https://github.com/theforeman/foreman/blob/develop/app/services/power_manager.rb#L2
  * [b] https://github.com/theforeman/foreman/blob/develop/app/views/hosts/select_multiple_power_state.html.erb#L7
* [2] "28.3 Chassis Control Command" at https://www.intel.la/content/dam/www/public/us/en/documents/specification-updates/ipmi-intelligent-platform-mgt-interface-spec-2nd-gen-v2-0-spec-update.pdf
* [3] "6.27.5.20 Reset Type" at https://www.dmtf.org/sites/default/files/standards/documents/DSP0268_2025.2.pdf                                                                               
* [4] When Foreman is instructed to run power operation via UI[1], what power action to call is determined by [c] and [d].
  * [c] https://github.com/theforeman/foreman/blob/develop/app/services/power_manager/bmc.rb#L17-L31
  * [d] https://github.com/theforeman/foreman/blob/develop/app/services/power_manager/base.rb#L18-L30
* [5] Mapping between "What power action Foreman determines to call[4]" & "What Foreman core actually calls against BMC proxy API" is determined by [e]
  * [e] https://github.com/theforeman/foreman/blob/develop/app/services/proxy_api/bmc.rb#L47-L66
* [6] Mapping between "power action instructed through BMC proxy API" and "which method of provider is called" is defined by [f]
  * [f] https://github.com/theforeman/smart-proxy/blob/develop/modules/bmc/bmc_api.rb#L163-L189
* [7] https://github.com/theforeman/smart-proxy/blob/develop/modules/bmc/ipmi.rb
* [8] https://github.com/theforeman/smart-proxy/blob/develop/modules/bmc/redfish.rb
* [PR10561] https://github.com/theforeman/foreman/pull/10561
* [PR916] https://github.com/theforeman/smart-proxy/pull/916
* [PR917] https://github.com/theforeman/smart-proxy/pull/917
* [PR919] https://github.com/theforeman/smart-proxy/pull/919

stejskalleos commented on app/services/power_manager/bmc.rb:
```diff
@@ -13,19 +13,40 @@ def ready?
 
     attr_reader :proxy
 
+    def has_poweraction_fix_3_17?
```
I don't know if mentioning the `3_17` version in the code is a good practice. 
What if we named it `power_action_v2` and put deprecation methods in the original `power_action`?

That way, we can in follow-up releases remove the deprecated method and keep only one.

evgeni replied:
This matches the new capability defined in the smart-proxy PRs, but I've been thinking to suggest two destinct capabilities `reboot` and `reset` that then can be used if available (falling back to `cycle` for setups not having the capability yet)?

vanou replied:
Thanks for comment.
Naming of capability is bit kind of preference. But whether to create 1 capability or 2 capabilities is important point.
If we create 2 separate capabilities, determination logic at [1] becomes bit complex, I think. So we made 1 capability because we expect all PRs[2][3][4][5] here will be merged at one Foreman release version, and not expecting cherry-picking part of these PRs at downstream.

[1] https://github.com/theforeman/foreman/pull/10561/files#diff-ff81ee231c9e9f24cd1650ca805a5fc1d6991ab0598234c6ea05764099420e3fR30-R31
[2] https://github.com/theforeman/foreman/pull/10561
[3] https://github.com/theforeman/smart-proxy/pull/916
[4] https://github.com/theforeman/smart-proxy/pull/917
[5] https://github.com/theforeman/smart-proxy/pull/919

stejskalleos replied:
We had a discussion with colleagues about putting the Foreman version into the function names, and the general consensus was that it's not a good practice.

Please rename it to something else. My suggestion is `power_action_v2`

evgeni commented:
Thanks @vanou!

Yeah, I agree, "soft off, on" for a soft reboot on IPMI is quite some work to get right, so I'd skip that for now too.

And huge thanks for the overview table and links, this helps a lot in understanding the facets of this issue.

One thing I noticed, `powerreset` is not defined in `Proxy::BMC::Base`, only in `Proxy::BMC::IPMI` (so calling this today on a Redfish machine would cause a NoMethod error, not an NotImplemented one), we probably can fix that as part of https://github.com/theforeman/smart-proxy/pull/916.

stejskalleos commented on app/services/power_manager/bmc.rb:
```diff
@@ -13,19 +13,40 @@ def ready?
+      nic = host.bmc_nic
+      nic&.subnet&.bmc&.has_capability?(:BMC, :poweraction_fix_3_17) || false
+    rescue => e
+      Foreman::Logging.logger('app').warn("Failed to determine BMC poweraction_fix_3_17 capability: #{e}")
```
Would it be beneficial to split the log message into two: `warn` & `debug`? I would personally leave the `e` for the debug, as it can pollute the logs quite easily. You know how verbose these messages can be.

vanou replied:
Thanks for comment.

Sorry. I'm not sure what message can be split into `warn` & `debug`. Currently, there is 1 log message, I think. And, maybe, there isn't many chances to raise error in this method, so we create only 1 `warn` here.

If possible, could you explain about this? I'm not familiar with this.

stejskalleos replied:
My idea was to split the messaging like this:
```ruby
Foreman::Logging.logger('app').warn("Failed to determine BMC poweraction_fix_3_17 capability")
Foreman::Logging.logger('app').error(e)
```

It's just a nitpick, but we can keep it in one message. 

However, in my opinion, it should be `.error`, since it's rescuing the `e`.

ekohl replied:
We also have `Foreman::Logging.exception(message, exception)`. Isn't that better to use here?

spesnova717 replied:
@stejskalleos 

Thank you for the suggestion and comments.  

We identified two cases that can happen in this method:  
1. A **subclass of ActiveRecordError** [1] occurs during the capability check.  
2. At `nic&.subnet&.bmc&.has_capability`, one of `nic`, `subnet`, or `bmc` can be `nil`, making the capability check invalid.  

The call `nic&.subnet&.bmc&.has_capability?(:BMC, :poweraction_fix_3_17)` mainly deals with an **ApplicationRecord (ActiveRecord) object**, so potential exceptions are expected to be subclasses of `ActiveRecordError`.  

Since either of these cases means the capability cannot be determined correctly, **continuing silently** (only output log and move forward, no interruption) could result in an unsafe power operation. Instead, the operation should be interrupted by raising an exception.  

With that reasoning, we will handle above cases as follows:  
- **Subclass of ActiveRecordError**  
  - Catch subclass of `ActiveRecordError`, then raise another exception so that it propagates, instead of silently continuing a potentially unsafe power operation.  
- **Nil cases (nil at `nic`, `subnet` or `bmc`)**  
  - If `nil` is detected, raise an exception since the capability check could not be performed correctly.  

Does this approach make sense to you?  

[1] https://api.rubyonrails.org/classes/ActiveRecord/Base.html

vanou commented:
> Yeah, I agree, "soft off, on" for a soft reboot on IPMI is quite some work to get right, so I'd skip that for now too.

Thanks you.

> And huge thanks for the overview table and links, this helps a lot in understanding the facets of this issue.

I'm glad to help understanding this complex issue.

>One thing I noticed, powerreset is not defined in Proxy::BMC::Base, only in Proxy::BMC::IPMI (so calling this today on a Redfish machine would cause a NoMethod error, not an NotImplemented one), we probably can fix that as part of https://github.com/theforeman/smart-proxy/pull/916.

Thanks. Actually, at https://github.com/theforeman/smart-proxy/pull/916, we define `powerreset` against `Proxy::BMC::Base` at [1].

[1] https://github.com/theforeman/smart-proxy/pull/916/files#diff-3d798e561908cf5ee7f4daf59f379deea3dc8abda25f7fd55c311a476d2aac87R49-R52

stejskalleos commented on app/services/power_manager/bmc.rb:
```diff
@@ -13,19 +13,40 @@ def ready?
+    def has_poweraction_fix_3_17?
+      nic = host.bmc_nic
+      nic&.subnet&.bmc&.has_capability?(:BMC, :poweraction_fix_3_17) || false
+    rescue => e
```
Can we be more specific about what errors we want to catch here?
This will catch _everything_, not sure if it's desired behavior.

spesnova717 replied:
@stejskalleos 

Thank you for the comments.
I responded to this in https://github.com/theforeman/foreman/pull/10561#discussion_r2351969734 .
Could you please check?

vanou commented:
Thanks @stejskalleos for review. I'll reply to your comment at Sep. 16 because Sep. 15 is national holiday.

evgeni commented:
Found another place where this is mixed up:
https://github.com/theforeman/foreman/blob/3d841a95caf6177e35862795158fcef95b29f441/app/controllers/api/v2/hosts_controller.rb#L245

The way the API doc is written implies `soft` and `reboot` are considered the same, and `cycle` and `reset` too.

evgeni commented:
And another one: https://github.com/theforeman/foreman/blob/3d841a95caf6177e35862795158fcef95b29f441/webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/constants.js#L6
(those constants are currently unused tho)

vanou commented:
> Found another place where this is mixed up:
> 
> https://github.com/theforeman/foreman/blob/3d841a95caf6177e35862795158fcef95b29f441/app/controllers/api/v2/hosts_controller.rb#L245
> 
> The way the API doc is written implies `soft` and `reboot` are considered the same, and `cycle` and `reset` too.

> And another one:
> 
> https://github.com/theforeman/foreman/blob/3d841a95caf6177e35862795158fcef95b29f441/webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/constants.js#L6
> 
> 
> (those constants are currently unused tho)

@evgeni 
Thanks for comment. We'll fix them in this PR too.

spesnova717 commented:
@stejskalleos @evgeni 

Thank you for your review.
I have updated the code based on the discussion.
I’d appreciate your review to confirm the changes.

stejskalleos commented on app/services/power_manager/bmc.rb:
```diff
@@ -13,19 +13,58 @@ def ready?
+        e,
+        level: :error
+      )
+      raise
```
Although a defensive approach is a good way to go, I don't think we need the `nil` check here.

Calling `host.power.action`:

* If the host doesn't have an associated `power`, the error will be catched `ERF42-9958 [Foreman::Exception]: Unknown power management support - can't continue`
* If the subnet doesn't have an BMC proxy, the error is `Foreman::BMCFeatureException: ERF88-9474 [Foreman::BMCFeatureException]: There is no proxy with BMC feature set up. Associate a BMC feature with a subnet.`

```suggestion
      smart_proxy = host.smart_proxies.with_features(:BMC).first
      
      return false unless smart_proxy      
      return smart_proxy.has_capability?(:BMC, :poweraction_fix_3_17)
```

What do you think?

vanou commented:
@stejskalleos 
Thanks for comment & suggestion. We discussed and agreed that your suggested code is reasonable.
So we modified this PR code based on your suggestion & consensus between you and your colleagues.

I'd appreciate your review. Thank you.

stejskalleos commented on test/unit/power_manager_test.rb:
```diff
@@ -82,4 +82,54 @@ def actions_list(host)
+    assert host.power.reset
+  end
+
+  test "should fallback to soft and cycle when poweraction_fix_3_17 is not supported" do
```
```suggestion
  test "should fallback to soft and cycle when power_action_v2 is not supported" do
```

stejskalleos commented on app/services/power_manager/bmc.rb:
```diff
@@ -13,19 +13,39 @@ def ready?
+      smart_proxy = host.smart_proxies.with_features(:BMC).first
+
+      return false unless smart_proxy
+      smart_proxy.has_capability?(:BMC, :poweraction_fix_3_17)
```
```suggestion
      smart_proxy.has_capability?(:BMC, :power_action_v2)
```

We should rename the method in the smart proxy as well (https://github.com/theforeman/smart-proxy/pull/916 & https://github.com/theforeman/smart-proxy/pull/917)

stejskalleos commented on test/unit/power_manager_test.rb:
```diff
@@ -82,4 +82,54 @@ def actions_list(host)
     actions.uniq
   end
+
+  test "should call reboot and reset directly when poweraction_fix_3_17 is supported" do
```
```suggestion
  test "should call reboot and reset directly when power_action_v2 is supported" do
```

stejskalleos reviewed (changes requested):
There are a few missing rename changes, and the smart proxy PRs also need to be updated; otherwise, the PR is ready for merge.

vanou commented:
@stejskalleos 
Thank you for comment on overlooked things. Now I fix things you told in code & commit message at this PR & PRs of smart-proxy.

Huge thanks for kind review and comment!

stejskalleos commented on app/services/power_manager/bmc.rb:
```diff
@@ -13,19 +13,39 @@ def ready?
     attr_reader :proxy
 
+    def power_action_v2
+      smart_proxy = host.smart_proxies.with_features(:BMC).first
```
Found a bug https://github.com/theforeman/foreman/pull/10697, the `smart_proxies` method does not include Proxies that are assigned as BMC only.

stejskalleos reviewed (approved):
:green_apple: **LGTM**

Tested two scenarios:
1. #10561 with https://github.com/theforeman/smart-proxy/pull/916/

```
[2025-09-23 13:35:00,183] INFO in main: System "202ca45a-09eb-4906-a0ef-926697668b44" power state set to "ForceRestart"
```

2. #10561 with https://github.com/theforeman/smart-proxy/pull/917

```
[2025-09-23 13:30:19,953] INFO in main: System "202ca45a-09eb-4906-a0ef-926697668b44" power state set to "GracefulRestart"
```

Both scenarios meet expectations and work well. I found one issue during testing, https://github.com/theforeman/foreman/pull/10697, which was not caused by these changes.

:green_apple: Works as expected
:green_apple: Code LGTM
:green_apple: Tests
:green_apple: CI is green

stejskalleos commented:
Thanks @spesnova717 @vanou @evgeni
